salt/transport/zeromq.py

```diff
@@ -736,6 +736,17 @@ def publish(self, load):
         if load['tgt_type'] == 'list':
             int_payload['topic_lst'] = load['tgt']
 
+        # If zmq_filtering is enabled, target matching has to happen master side
+        if self.opts['zmq_filtering']:
```

Comment:
This won't quite work. ckminions (on the master) does a best effort guess of who the master thinks is the target. The issue is for things that are minion-side targeted (data, module, etc.) them master has no idea about them. This gets especially sticky with syndics, as the top level master doesn't know about the minions behind a syndic either. Similar issues in #18358 for more context.

Basically at this point we don't have a reliable way of targeting minions on the master-side for most target types. So, if you want to implement this I believe we are limited to master-side data (minion-id, pillar, maybe grains).